File: examples/cxx/ping_pong/Sender.cxx
#include "Sender.hpp"
#include "PingPongTask.hpp"
#include <MsgHost.hpp>
#include <HostNotFoundException.hpp>
#include <Msg.hpp>

#include <Msg.hpp>
using namespace std;

const double commSizeLat = 1;
const double commSizeBw = 100000000;

MSG_IMPLEMENT_DYNAMIC(Sender, Process)

int Sender::main(int argc, char** argv)
{
	info("Hello");
	
	int hostCount = argc;
	
	info(TEXT_("Host count : ") + TEXT_(hostCount));
	
	Host* hostTable = new Host[hostCount]; 
	double time;
	double computeDuration = 0;
	PingPongTask* task;
	
	for(int pos = 0; pos < argc ; pos++) 
	{
		try 
		{
			hostTable[pos] = Host::getByName(argv[pos]);
		} 
		catch(HostNotFoundException e) 
		{
			error(TEXT_(e.toString()) + TEXT_(". Stopping Now!"));
			exit(1);
		}
	}
	
	for (int pos = 0; pos < hostCount; pos++) 
	{ 
		time = getClock(); 
		
		info(TEXT_("Sender time : ") + TEXT_(time));
	
		task = new PingPongTask("no name",computeDuration,commSizeLat);
		task->setTime(time);
	
		hostTable[pos].put(0,task);
	} 
	
	info("Goodbye!");

	
	delete[] hostTable;

	delete this;

	return 0;
    
}

File: examples/java/tracing/Sender.java
/* Copyright (c) 2006-2014, 2016. The SimGrid Team.
 * All rights reserved.                                                     */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

package tracing;
import org.simgrid.msg.Msg;
import org.simgrid.msg.Host;
import org.simgrid.msg.Task;
import org.simgrid.msg.Process;
import org.simgrid.msg.MsgException;
import org.simgrid.msg.HostNotFoundException;
import org.simgrid.trace.Trace;

public class Sender extends Process {
  private final double commSizeLat = 1;
  private final double commSizeBw = 100000000;

  public Sender(Host host, String name, String[] args) {
    super(host,name,args);
  }

  public void main(String[] args) throws MsgException {
    Msg.info("hello !"); 
    Trace.hostPushState (getHost().getName(), "PM_STATE", "sendingPing");

    int hostCount = args.length;
    Msg.info("host count: " + hostCount);
    String mailboxes[] = new String[hostCount]; 
    double time;
    double computeDuration = 0;
    PingPongTask ping, pong;

    for(int pos = 0; pos < args.length ; pos++) {
      try {
        mailboxes[pos] = Host.getByName(args[pos]).getName();
      } catch (HostNotFoundException e) {
        Msg.info("Invalid deployment file: " + e.toString());
        System.exit(1);
      }
    }

    for (int pos = 0; pos < hostCount; pos++) { 
      time = Msg.getClock(); 
      Msg.info("sender time: " + time);
      ping = new PingPongTask("no name",computeDuration,commSizeLat);
      ping.setTime(time);
      ping.send(mailboxes[pos]);

      Trace.hostPushState (getHost().getName(), "PM_STATE", "waitingPong");
      pong = (PingPongTask)Task.receive(getHost().getName());
      double timeGot = Msg.getClock();
      double timeSent = ping.getTime();
      double communicationTime=0;

      Msg.info("Got at time "+ timeGot);
      Msg.info("Was sent at time "+timeSent);
      time=timeSent;

      communicationTime=timeGot - time;
      Msg.info("Communication time : " + communicationTime);

      Msg.info(" --- bw "+ commSizeBw/communicationTime + " ----");

      /* Pop the last state (going back to sending ping) */  
      Trace.hostPopState (getHost().getName(), "PM_STATE");
    }

    /* Pop the sendingPong state */  
    Trace.hostPopState (getHost().getName(), "PM_STATE");
    Msg.info("goodbye!");
  }
}
